src/devices_models/devices/thermal_generation.jl

```diff
@@ -540,6 +540,85 @@ function duration_initial_condition!(
     return
 end
 
+############################ Auxiliary Variables Calculation ################################
+function calculate_aux_variable_value!(
+    optimization_container::OptimizationContainer,
+    key::AuxVarKey{TimeDurationOn, T},
+    ::PSY.System,
+) where {T <: PSY.ThermalGen}
+    on_var_results = get_variable(optimization_container, OnVariable, T)
+    aux_var_container = get_aux_variables(optimization_container)[key]
+    ini_cond = get_initial_conditions(optimization_container, InitialTimeDurationOn, T)
+
+    time_steps = model_time_steps(optimization_container)
+    resolution = model_resolution(optimization_container)
+    minutes_per_period = Dates.value(Dates.Minute(resolution))
+
+    for ix in eachindex(JuMP.axes(aux_var_container)[1])
+        @assert JuMP.axes(aux_var_container)[1][ix] == JuMP.axes(on_var_results)[1][ix]
+        @assert JuMP.axes(aux_var_container)[1][ix] == get_device_name(ini_cond[ix])
+        on_var = JuMP.value.(on_var_results.data[ix, :])
+        ini_cond_value = get_condition(ini_cond[ix])
+        aux_var_container.data[ix, :] .= ini_cond_value
+        sum_on_var = sum(on_var)
+        if sum_on_var == time_steps[end] # Unit was always on
+            aux_var_container.data[ix, :] += time_steps * minutes_per_period
+        elseif sum_on_var == 0.0 # Unit was always off
+            aux_var_container.data[ix, :] .= 0.0
+        else
+            previous_condition = ini_cond_value
+            for (t, v) in enumerate(on_var)
+                if v < 0.99 # Unit turn off
+                    aux_var_container.data[ix, t] = 0.0
+                elseif isapprox(v, 1.0) # Unit is on
+                    aux_var_container.data[ix, t] = previous_condition + 1.0
```

Comment:
nit: extra dereference of this container. You could use a temporary variable and access the container once. Same in the other function.